internal/client/v1/client/discoverer/discover.go

```diff
@@ -218,111 +218,139 @@ func (c *client) dnsDiscovery(ctx context.Context, ech chan<- error) (addrs []st
+}
+
+func (c *client) discoverAddrs(ctx context.Context, nodes *payload.Info_Nodes, ech chan<- error) (addrs []string, err error) {
+	maxPodLen := 0
+	podLength := 0
+	for _, node := range nodes.GetNodes() {
+		l := len(node.GetPods().GetPods())
+		podLength += l
+		if l > maxPodLen {
+			maxPodLen = l
 		}
-		addrs := make([]string, 0, podLength)
-		for i := 0; i < maxPodLen; i++ {
-			for _, node := range nodes.GetNodes() {
-				select {
-				case <-ictx.Done():
-					return nil, ictx.Err()
-				default:
-					if node != nil && node.GetPods() != nil {
-						pods := node.GetPods().GetPods()
-						if i < len(pods) {
-							addr := net.JoinHostPort(pods[i].GetIp(), uint16(c.port))
-							if err = c.connect(ctx, addr); err != nil {
-								err = errors.ErrAddrCouldNotDiscover(err, addr)
-								select {
-								case <-ictx.Done():
-									return nil, ictx.Err()
-								case ech <- err:
-								}
-								err = nil
-							} else {
-								if c.autoconn {
-									cur.Store(addr, struct{}{})
-								}
-								addrs = append(addrs, addr)
+	}
+	addrs = make([]string, 0, podLength)
+	for i := 0; i < maxPodLen; i++ {
+		for _, node := range nodes.GetNodes() {
+			select {
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			default:
+				if node != nil && node.GetPods() != nil {
+					pods := node.GetPods().GetPods()
+					if i < len(pods) {
+						addr := net.JoinHostPort(pods[i].GetIp(), uint16(c.port))
+						if err = c.connect(ctx, addr); err != nil {
+							err = errors.ErrAddrCouldNotDiscover(err, addr)
+							select {
+							case <-ctx.Done():
+								return nil, ctx.Err()
+							case ech <- err:
 							}
-							break
+							err = nil
+						} else {
+							addrs = append(addrs, addr)
 						}
+						break
 					}
 				}
 			}
 		}
-		connected = addrs
-		if len(connected) == 0 {
-			log.Warn("connected addr is zero")
-			cur = sync.Map{}
-			return nil, errors.ErrAddrCouldNotDiscover(err, c.dns)
-		}
-		if c.onDiscover != nil {
-			err = c.onDiscover(ctx, c, connected)
-			if err != nil {
-				cur = sync.Map{}
-				return nil, err
-			}
-		}
-		return nil, nil
-	}); err != nil {
-		log.Warn("failed to discover addrs from discoverer API, trying to discover from dns...\t" + err.Error())
-		connected, err = c.dnsDiscovery(ctx, ech)
-		if err != nil {
-			return err
-		}
-		if c.autoconn {
-			cur = sync.Map{}
-			for _, addr := range connected {
-				cur.Store(addr, struct{}{})
-			}
-		}
 	}
+	return addrs, nil
+}
 
-	oldAddrs := c.GetAddrs(ctx)
-
-	c.addrs.Store(connected)
+func (c *client) disconnectOldAddrs(ctx context.Context, oldAddrs, connectedAddrs []string, ech chan<- error) (err error) {
+	if !c.autoconn {
+		return nil
+	}
+	var cur sync.Map
+	for _, addr := range connectedAddrs {
+		cur.Store(addr, struct{}{})
+	}
 
 	for _, old := range oldAddrs {
 		_, ok := cur.Load(old)
 		if !ok {
 			c.eg.Go(safety.RecoverFunc(func() error {
-				err = c.disconnect(ctx, old)
+				err := c.disconnect(ctx, old)
```

Comment:
can you describe about this?
I thought we can reuse returning definition.

Reply:
Thank you for your comment.
I have fixed: https://github.com/vdaas/vald/pull/1706/commits/45057d7b5c5bdb82b61c570e511f6d81bb475e54